File: common-test/src/main/java/io/quartic/common/test/rx/Interceptor.java
package io.quartic.common.test.rx;

import com.google.common.collect.ImmutableList;
import rx.Observable;
import rx.Observable.Transformer;

import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static com.google.common.collect.Lists.newArrayList;
import static java.util.Collections.synchronizedList;

public final class Interceptor<T> implements Transformer<T, T> {
    private final AtomicInteger subscribeCount = new AtomicInteger(0);
    private final AtomicInteger unsubscribeCount = new AtomicInteger(0);
    private final List<T> values = synchronizedList(newArrayList());

    public static <T> Interceptor<T> create() {
        return new Interceptor<>();
    }

    public boolean subscribed() {
        return subscribeCount.get() > 0;
    }

    public boolean unsubscribed() {
        return unsubscribeCount.get() > 0;
    }

    public int subscribeCount() {
        return subscribeCount.get();
    }

    public int unsubscribeCount() {
        return unsubscribeCount.get();
    }

    public List<T> values() {
        return ImmutableList.copyOf(values);
    }

    @Override
    public Observable<T> call(Observable<T> observable) {
        return observable
                .doOnNext(values::add)  // Note this will be weird if we have multiple subscribers
                .doOnSubscribe(subscribeCount::getAndIncrement)
                .doOnUnsubscribe(unsubscribeCount::getAndIncrement);
    }
}

File: common-test/src/main/kotlin/io/quartic/common/test/rx/Interceptor.kt
package io.quartic.common.test.rx

import rx.Observable
import rx.Observable.Transformer
import java.util.concurrent.atomic.AtomicInteger

class Interceptor<T> : Transformer<T, T> {
    private val _subscribeCount = AtomicInteger(0)
    private val _unsubscribeCount = AtomicInteger(0)
    private val _values = mutableListOf<T>()

    val subscribed: Boolean get() = subscribeCount > 0
    val unsubscribed: Boolean get() = unsubscribeCount > 0
    val subscribeCount: Int get() = _subscribeCount.get()
    val unsubscribeCount: Int get() = _unsubscribeCount.get()
    val values: List<T> get() = _values.toList()

    override fun call(observable: Observable<T>): Observable<T> {
        return observable
                .doOnNext({ synchronized(_values) { _values.add(it) } })  // Note this will be weird if we have multiple subscribers
                .doOnSubscribe({ _subscribeCount.incrementAndGet() })
                .doOnUnsubscribe({ _unsubscribeCount.incrementAndGet() })
    }
}